// vuetify/es5/components/VResponsive/VResponsive.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("../../../src/components/VResponsive/VResponsive.sass");

var _measurable = _interopRequireDefault(require("../../mixins/measurable"));

var _mixins = _interopRequireDefault(require("../../util/mixins"));

var _helpers = require("../../util/helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Mixins
// Utilities

/* @vue/component */
var _default = (0, _mixins.default)(_measurable.default).extend({
  name: 'v-responsive',
  props: {
    aspectRatio: [String, Number]
  },
  computed: {
    computedAspectRatio: function computedAspectRatio() {
      var aspectRatio = Number(this.aspectRatio);
      return !isNaN(aspectRatio) ? 1 / aspectRatio : undefined;
    },
    computedAspectRatioPercent: function computedAspectRatioPercent() {
      if (this.computedAspectRatio == null) return undefined;
      return this.computedAspectRatio * 100;
    },
    contentAspectStyle: function contentAspectStyle() {
      if (this.computedAspectRatioPercent == null) return undefined;
      return {
        marginTop: (0, _helpers.convertToUnit)(-1 * this.computedAspectRatioPercent, '%')
      };
    },
    sizerAspectStyle: function sizerAspectStyle() {
      if (this.computedAspectRatioPercent == null) return undefined;
      return {
        paddingTop: (0, _helpers.convertToUnit)(this.computedAspectRatioPercent, '%')
      };
    }
  },
  methods: {
    genContent: function genContent() {
      return this.$createElement('div', {
        staticClass: 'v-responsive__sizer',
        style: this.sizerAspectStyle
      }, [this.$createElement('div', {
        staticClass: 'v-responsive__content',
        style: this.contentAspectStyle
      }, this.$slots.default)]);
    }
  },
  render: function render(h) {
    return h('div', {
      staticClass: 'v-responsive',
      style: this.measurableStyles,
      on: this.$listeners
    }, [this.genContent()]);
  }
});

exports.default = _default;
